[gcp] Add metadata cache

gpop63 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Overview

This PR adds a cache for existing metadata collectors:
- `compute`
- `dataproc`
- `redis`
- `cloudsql`

Two new config variables are added: `metadata_cache` (bool) to enable the cache and `metadata_cache_refresh_period` (period) to decide the refresh interval.

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Beats.
-->

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #43959

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

mrg-elastic committed:
add metadata cache

mrg-elastic committed:
Merge remote-tracking branch 'upstream/main' into gcp/add_metadata_cache

mrg-elastic committed:
add dataproc metadata cache

mrg-elastic committed:
add changelog entry

mrg-elastic committed:
update docs

mrg-elastic committed:
fix pr id

mrg-elastic committed:
add license

mrg-elastic committed:
fix imports

belimawr reviewed (approved):
The files owned by the data-plane (the configuration files) look good. Approving the change on those files.

mrg-elastic committed:
add generics for cache getters

mrg-elastic committed:
improve code with generics

gpop63 commented on x-pack/metricbeat/module/gcp/metadata_cache.go:
```diff
@@ -0,0 +1,356 @@
+		cacheCopy[k] = v
+	}
+	return cacheCopy
+}
```
Good idea, I added generics to make the code less redundant and more maintainable [ef1a424](https://github.com/elastic/beats/pull/44432/commits/ef1a4244807cf5f60c18b98bbfc8aa6a77045d83) WDYT?

devamanv replied:
Looks good! Thanks for making the change.

mrg-elastic committed:
lock in update methods

mrg-elastic committed:
add tests

mrg-elastic committed:
fix impoirts

Kavindu-Dodan commented on x-pack/metricbeat/module/gcp/metrics/compute/metadata.go:
```diff
@@ -180,44 +198,49 @@ func (s *metadataCollector) instanceZone(ts *monitoringpb.TimeSeries) string {
 		}
 	}
+
+	s.logger.Infof("Compute AggregatedList finished in %s. Fetched %d instances across %d pages.", time.Since(start), instanceCount, pageCount)
```
nit : should this be `info` level ? Prior to the refactoring, it was a defer based `debug` level log isn't it ?

gpop63 replied:
good catch thanks!

ishleenk17 committed:
Merge branch 'main' into gcp/add_metadata_cache

Kavindu-Dodan commented on x-pack/metricbeat/module/gcp/metadata_cache.go:
```diff
@@ -0,0 +1,238 @@
+func getCache[T any](r *CacheRegistry, cache *Cache[T]) map[string]T {
+	r.dataMutex.RLock()
+	defer r.dataMutex.RUnlock()
+	return cacheCopy(cache.data)
```
> Do you think instead of this I should create a method that gets a specific instance by id?

If we only need access by ID, then I think we can avoid the copy and make sure we provide an immutable access to the value (note - Value itself can be a pointer so it's sort of not guaranteed to be immutable. But that's too much of an edge case so I wouldn't worry :) )

So yeah, I think it's a good improvement :)

gpop63 replied:
Addressed in d6eb27763ec89b24f62ae13ac2095c46e5616795

Kavindu-Dodan replied:
That's fast and looks good :)

mrg-elastic committed:
add getters by id
add fetchFunc exponential backoff with max 3 retries

mrg-elastic committed:
update tests

gpop63 commented on x-pack/metricbeat/module/gcp/_meta/docs.asciidoc:
```diff
@@ -45,6 +45,12 @@ and metadata information from metricsets and fetch metrics only. At the moment,
 * *collect_dataproc_user_labels*: (`true`/`false` default `false`) Retrieve additional
 user-defined labels from Dataproc clusters.
 
+* *metadata_cache*: (`true`/`false` default `false`) Enable caching of metadata.
```
I don't really have a strong opinion on this, all of them look good to me

mrg-elastic committed:
use debug level instead of info

mrg-elastic committed:
make update

mrg-elastic committed:
fix golangci-lint

Kavindu-Dodan commented on x-pack/metricbeat/module/gcp/metadata_cache.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// UpdateComputeCache appends the provided map to the Compute instance cache.
+func (r *CacheRegistry) UpdateComputeCache(update map[string]*computepb.Instance) {
```
nit : regarding updates, seems they are not used. Is there any plan to use them or do you see the need ?

gpop63 replied:
initially I thought I will use them in each metadata collector but now that is being done in `ensureCacheFresh`, I will remove them

Kavindu-Dodan reviewed (approved):
Only have a nit otherwise looks okay. Approving as a code owner :)

ishleenk17 reviewed (approved):
Looks good!

ishleenk17 commented on CHANGELOG.next.asciidoc:
```diff
@@ -519,6 +519,7 @@ otherwise no tag is added. {issue}42208[42208] {pull}42403[42403]
 - Add new metrics to vSphere Virtual Machine dataset (CPU usage percentage, disk average usage, disk read/write rate, number of disk reads/writes, memory usage percentage). {pull}44205[44205]
 - Added checks for the Resty response object in all Meraki module API calls to ensure proper handling of nil responses. {pull}44193[44193]
 - Add latency config option to Azure Monitor module. {pull}44366[44366]
+- Add GCP cache for metadata collectors. {pull}44432[44432]
```
Now the relase.md has the 9.3.0 section. LEts add it there only

mrg-elastic committed:
move changelog entry

mrg-elastic committed:
remove unused update methods
update tests

colleenmcginnis commented on docs/release-notes/index.md:
```diff
@@ -34,6 +34,7 @@ To check for security updates, go to [Security announcements for the Elastic sta
 
 - Add checks for the Resty response object in all Meraki module API calls to ensure proper handling of nil responses. [44193]({{beats-pull}}44193)
 - Add a latency configuration option to the Azure Monitor module. [44366]({{beats-pull}}44366)
+- Add GCP cache for metadata collectors. [44432]({{beats-pull}}44432)
```
Did this change go into 9.0.2 or is it going in a future release? This line is in the 9.0.2 section of the release notes. If the fix didn't make it into 9.0.2, you should add it to the appropriate 9.0.x section instead (more towards the top of this file).

gpop63 replied:
it's going in `9.1.0` actually, I don't see that version present in the file where should it be added?

colleenmcginnis replied:
Yes exactly!

mrg-elastic committed:
update gcp reference doc

colleenmcginnis commented on docs/reference/metricbeat/metricbeat-module-gcp.md:
```diff
@@ -36,6 +36,9 @@ This is a list of the possible module parameters you can tune:
 * **exclude_labels**: (`true`/`false` default `false`) Do not extract extra labels and metadata information from metricsets and fetch metrics only. At the moment, **labels and metadata extraction is only supported** in `compute` metricset.
 * **period**: A single time duration specified for this module collection frequency.
 * **endpoint**: A custom endpoint to use for the GCP API calls. If not specified, the default endpoint will be used.
+* **metadata_cache**: (`true`/`false` default `false`) Enable caching of metadata. If set to true, metadata will be cached to improve performance.
+* **metadata_cache_refresh_period**: A duration specifying how often the cached metadata should be refreshed (e.g., `5m`, `1h`).
```
Going forward we should be labeling new features with the version number it was added in. I'm working on some updated contributing guidelines, but for now this should suffice:

```suggestion
* **metadata_cache**: (`true`/`false` default `false`) Enable caching of metadata. If set to true, metadata will be cached to improve performance. {applies_to}`product: ga 9.1.0`
* **metadata_cache_refresh_period**: A duration specifying how often the cached metadata should be refreshed (e.g., `5m`, `1h`). {applies_to}`product: ga 9.1.0`
```

We're actively working on improving the rendered version badges, but for now this is what it will look like:

<img width="841" alt="image" src="https://github.com/user-attachments/assets/ab273249-2b21-4fac-84c9-b57415578dbe" />

Commit:
Update docs/reference/metricbeat/metricbeat-module-gcp.md

Co-authored-by: Colleen McGinnis <colleen.j.mcginnis@gmail.com>

mrg-elastic committed:
fix changelog

mrg-elastic committed:
remove cacheRegistry nil check
use old calls when cache is disabled
keep old maps to store instances without cache

mrg-elastic committed:
fix golangci-lint

tommyers-elastic committed:
add retry helper for libbeat backoff implementation

mrg-elastic committed:
fix golangci-lint

mrg-elastic committed:
simplify cache
refactor generic cache struct
move cache refresh logic into Cache[T].EsureFresh method
remove backoff retry logic
remove complex double locking
remove redundant instance call from metadata collectors

mrg-elastic committed:
make update

gpop63 commented:
@tommyers-elastic suggested a better and simpler approach. Some of the changes:

- the previous cache logic was a bit complex with 2 mutexes and double locking; cache is now more self contained
- moved retry logic to the client (metadata collectors)
- switched to elastic's own retry library instead of adding a new dependency
- metadata for each service (compute, redis, etc.) is now populated during `NewMetadataService`, this removes the need for `instance` methods in collectors
- got rid of redundant metadata maps that store data in each service, now storing everything directly in the cache registry’s service specific maps

mrg-elastic committed:
remove redundant metadata cache check

gpop63 commented:
/test

mrg-elastic committed:
extra check for potential windows clock issue

tommyers-elastic commented on x-pack/metricbeat/module/gcp/metadata_cache.go:
```diff
@@ -0,0 +1,97 @@
+
+func (c *Cache[T]) isExpired() bool {
+	// A zero or negative interval means the cache is always considered expired.
+	if c.refreshInterval <= 0 {
```
this is fine, but the check below also covers this case since the result of `time.Since` will always be positive.

gpop63 replied:
without this check the Windows packaging tests always failed, it's some clock issue specific to windows

tommyers-elastic replied:
weird. but ok, it's not hurting anything.

tommyers-elastic commented on x-pack/metricbeat/module/gcp/metrics/cloudsql/metadata.go:
```diff
@@ -137,16 +153,17 @@ func (s *metadataCollector) instanceRegion(ts *monitoringpb.TimeSeries) string {
 	}
 
+	instance, ok := s.instanceCache.Get(instanceID)
+	if !ok {
```
do we need both the `!ok` check and the `nil` check?

tommyers-elastic commented on x-pack/metricbeat/module/gcp/metrics/compute/metadata.go:
```diff
@@ -180,44 +185,49 @@ func (s *metadataCollector) instanceZone(ts *monitoringpb.TimeSeries) string {
+		pageCount++
 
 		instances := instancesScopedListPair.Value.GetInstances()
+		if instances == nil {
```
nit: this isn't required, `range` over a nil map is a noop anyway

tommyers-elastic commented on x-pack/metricbeat/module/gcp/metrics/dataproc/metadata.go:
```diff
@@ -105,16 +120,17 @@ func (s *metadataCollector) Metadata(ctx context.Context, resp *monitoringpb.Tim
 	}
 
+	cluster, ok := s.instanceCache.Get(instanceID)
+	if !ok {
```
again, do we need both checks here? (i think just `!ok` is enough?)

tommyers-elastic commented on x-pack/metricbeat/module/gcp/metrics/redis/metadata.go:
```diff
@@ -102,16 +118,17 @@ func (s *metadataCollector) Metadata(ctx context.Context, resp *monitoringpb.Tim
 	}
 
+	instance, ok := s.instanceCache.Get(instanceID)
+	if !ok {
```
and here

tommyers-elastic reviewed (approved):
LGTM - thanks gabriel

mrg-elastic committed:
remove redundant nil checks

colleenmcginnis reviewed (approved):
✨

daniela-elastic commented on docs/reference/metricbeat/metricbeat-module-gcp.md:
```diff
@@ -36,6 +36,9 @@ This is a list of the possible module parameters you can tune:
 * **exclude_labels**: (`true`/`false` default `false`) Do not extract extra labels and metadata information from metricsets and fetch metrics only. At the moment, **labels and metadata extraction is only supported** in `compute` metricset.
 * **period**: A single time duration specified for this module collection frequency.
 * **endpoint**: A custom endpoint to use for the GCP API calls. If not specified, the default endpoint will be used.
+* **metadata_cache**: (`true`/`false` default `false`) Enable caching of metadata. If set to true, metadata will be cached to improve performance. {applies_to}`product: ga 9.1.0`
```
As discussed during the weekly call, let's make sure this doesn't impact new resources including ephemeral resources (that come and go)

ishleenk17 replied:
@gpop63 : I am in favor of handling that as enhance,net. We can create ticket and pick up in next sprint.
Lets bring this to closure.
cc: @lalit-satapathy

mrg-elastic committed:
enable cache by default

mrg-elastic committed:
update docs

mrg-elastic committed:
fix docs

mrg-elastic committed:
improve docs

mrg-elastic committed:
Revert "improve docs"

This reverts commit 6592a155a9f70b9d5785ef2728afb00c0e975964.

mrg-elastic committed:
Revert "fix docs"

This reverts commit e610dfb35bbac79b76a74ed59e26c5a2d28f0579.

mrg-elastic committed:
Revert "update docs"

This reverts commit de80b30dedd47d81c26e360eb4a3fa1bb3a20d62.

mrg-elastic committed:
Revert "enable cache by default"

This reverts commit d2f895cbf2a50fa2b27a6454edc50ff381ffa92c.

mrg-elastic committed:
document cache behavior
